src/WCSimWCAddDarkNoise.cc

```diff
@@ -0,0 +1,254 @@
+#include "WCSimPmtInfo.hh"
+#include "WCSimDarkRateMessenger.hh"
+
+#include <vector>
+#include <utility>
+// for memset
+#include <cstring>
+#include <iostream>
+
+#ifndef WCSIMWCADDDARKNOISE_VERBOSE
+//#define WCSIMWCADDDARKNOISE_VERBOSE
+#endif
+
+WCSimWCAddDarkNoise::WCSimWCAddDarkNoise(G4String name,
+                                   WCSimDetectorConstruction* myDetector)
+  :G4VDigitizerModule(name)
+{
+  this->myDetector = myDetector;
+  DigiHitMap.clear();
+  DarkRateMessenger = new WCSimDarkRateMessenger(this);
+}
+
+WCSimWCAddDarkNoise::~WCSimWCAddDarkNoise(){
+  DarkRateMessenger = 0;
+}
+
+void WCSimWCAddDarkNoise::AddDarkNoise(){
+  G4DigiManager* DigiMan = G4DigiManager::GetDMpointer();
+  // Get the PMT collection ID                                                                                                                                             
+  G4int WCHCID = DigiMan->GetDigiCollectionID("WCRawPMTSignalCollection");
+  // Get the PMT Digits collection                                                                                                                                        
+  WCSimWCDigitsCollection* WCHCPMT =
+    (WCSimWCDigitsCollection*)(DigiMan->GetDigiCollection(WCHCID));
+  
+  if (WCHCPMT && this->PMTDarkRate>1E-307) {
+    //Determine ranges for adding noise
+    if(DarkMode == 1)
+      FindDarkNoiseRanges(WCHCPMT,this->DarkWindow);
+    else if(DarkMode == 0) {
+      result.push_back(std::pair<float,float>(DarkLow,DarkHigh));
+    }
+    //Call routine to add dark noise here.
+    //loop over pairs which represent ranges.
+    //Add noise to those ranges
+    std::vector<std::pair<float, float> >::iterator it2 = result.begin();
+    while (it2 != result.end()){
+      AddDarkNoiseBeforeDigi(WCHCPMT,it2->first,it2->second);
+      it2++;
+    }
+  }
+}
+
+void WCSimWCAddDarkNoise::AddDarkNoiseBeforeDigi(WCSimWCDigitsCollection* WCHCPMT, float num1 ,float num2) {
+    // Introduces dark noise into each PMT during an event window
+    // This won't introduce noise only events, and isn't written
+    // to handle different rates for each PMT (although this shouldn't
+    // be too difficult to add at a later time)
+    // 
+    // Added by: Morgan Askins (maskins@ucdavis.edu)
+
+    G4int number_entries = WCHCPMT->entries();
+    const G4int number_pmts = myDetector->GetTotalNumPmts();
+    int *PMTindex = new int [number_pmts+1];
+    //   int PMTindex[number_pmts];
+    //initialize PMTindex
+    for (int l=0; l<number_pmts+1; l++){
+      PMTindex[l] =0;
+    }
+    int num_hit_b4=0;
+    //int num_hit_after=0;
+    //    std::cout<<"entries before "<<WCHCPMT->entries()<<"\n";
+    //Set up proper indices for tubes which have already been hit
+    for (int g=0; g<number_entries; g++){
+      G4int tube = (*WCHCPMT)[g]->GetTubeID();
+      //std::cout<<"totalpe "<<tube<<" "<<(*WCHCPMT)[g]->GetTotalPe()<<"\n";
+      for (int gp=0; gp<(*WCHCPMT)[g]->GetTotalPe(); gp++){
```

Comment:
Isn;t this for loop just the same as 
 PMTindex[tube]+=(*WCHCPMT)[g]->GetTotalPe();
 num_hit_b4+=(*WCHCPMT)[g]->GetTotalPe();
???